File: Problems/No.46_Permutations/permutations.cpp
class Solution {
public:
    
    // find all permutations using backtracking
    void solve(vector<int> &permutation, vector<vector<int>> &all_permutations, int index, int max) {
        
        if (index >= max) {
            
            all_permutations.push_back(permutation);
            return;
            
        }
        else {
            
            for (int i=index; i<max; i++) {
                
                swap(permutation[index], permutation[i]);
                solve(permutation, all_permutations, index+1, max);
                swap(permutation[index], permutation[i]);
                
            }
            
        }
        
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        
        vector<vector<int>> all_permutations;   // parameter
        solve(nums, all_permutations, 0, nums.size());  // start compute, using backtracking
        return all_permutations;
        
    }
};

File: Problems/No.46_Permutations/permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        # find all permutations using backtracking
        def solve(nums, permutation):
            if (len(permutation) == length):
                all_permutations.append(permutation)
            for i in range(len(nums)):
                solve(nums[:i]+nums[i+1:], permutation + [nums[i]])
                
        # parameters
        all_permutations = []  
        length = len(nums)
        
        # start compute
        solve (nums, [])
        return all_permutations
